### DISCUSSION TITLE
Roadmap Ideas...

### DISCUSSION THREAD

**Discussion by ebonow** [JedWatson/react-select#4376] [General]
I wanted to start a somewhat public facing checklist of areas I think need to be addressed. Perhaps this is a starting point of creating some other milestones. If anyone has any other ideas or support, please feel free to reply and I can append to this list. 

At some point I will likely begin to assign some sense of prioritization and milestones should collaboration team agree on prioritization.

- [x] React 17 - *Already planned for v4*
  - ~dependency~
  - ~life cycle methods~
  - ~hooks~
- Menu Management
  - [ ] Encapsulate menu events to an `onMenuChange` prop to include an action api to include events like:
    - willOpen/opening
    - didOpen/opened
    - willClose/closing
    - didClose/closed
  - [ ] Consider adding menuListRef to `onMenuOpen` prop   
  - [ ] Force async options reload using `cacheOptions` prop
  - [ ] IntersectionObserver for Menu
    - [Performance gains](https://itnext.io/1v1-scroll-listener-vs-intersection-observers-469a26ab9eb6 ) 
    - [ ] Responsive layout positioning when `menuPlacement` set to auto
    - [ ] create prop `onScroll` to replace `onScrollTop/Bottom`
``` jsx
       const onScroll = ({ scrollHeight, offset, isTop, isBottom }) => {}

```
  - [ ] Allow touchStart/mouseDown outside menuList (menu headers/footers)
  - [ ] Allow onClick handlers inside menu (currently swallowed by stopPropagation)
  - [ ] setFocusedOption method (examples below)
    - Expose method as parameter `onMenuOpen` or as suggested above `onMenuChange({ action: "opening" })`
```js
    - setFocusedOption() // default focus behavior
    - setFocusedOption(null) // no focused option
    - setFocusedOption(0) // first option
    - setFocusedOption(-1) // last option
    - setFocusedOption(o => o.label.indexOf(inputValue) === 0) // first match based on predicate
```
  - [ ] method to open/close without controlled input (which has side effect of bypassing onMenuOpen/Close callbacks)
  - [ ] animatedComponent
  - [ ] encapsulate menu logic outside of Select.js
    - consider possibility of Menu as standalone component
  - [ ] Apply "focus" styling using `:hover` and className (for keyboard) to improve scrolling performance by removing onMouseEnter/onMouseLeave DOM event binding
  - [ ] use cursor: pointer for options
  - [ ] consider default for captureScrollEvent to false or include { passive: false } to remove chrome warnings
- Accessibility
  - [ ] aria message customization *Already planned for v3.3*
  - [ ] aria message encapsulation into own component/utility
  - [ ] allow aria messaging to be disabled in favor of aria attributes
  - [ ] apply best guess aria attributes when aria messaging config is disabled 
  - [ ] Investigate aria messaging when focusing on grouped option (1 of 2 _or _1 of 13)
- [ ] Fix nested select references when applying async/creatable
- [ ] new createFilterOptions
  - [ ] maxOptions
  - [ ] sorting
- [ ] Add `selectProps` to InputProps style api
- [ ] Allow for replacement of custom Input component
- [ ] Validation and Required
- [ ] Investigate use of `trigger` method for testing or replicating functionality in custom components
- [ ] `disableGroups` prop to disable grouping or introduce new `groups` prop in case options includes an option named "options" but it isn't a group

**Comment by dalyhabit** [comment]
- Allow selectable Groups (when user clicks on a GroupHeading, every option within the Group is selected)
- Allow custom components within the Menu to be focusable / keyboard accessible (currently, focusing on anything other than an Option causes the menu to blur)
- Add a `MenuFooter` prop or at least allow custom menu footers to be keyboard accessible
